src/kernel/mutex.c

```diff
@@ -0,0 +1,148 @@
+#include <kernel.h>
+
+//#define MUTEX_DEBUG
+#ifdef MUTEX_DEBUG
+#define mutex_debug(x, ...) do {log_printf(" MTX", "%s: " x, __func__, ##__VA_ARGS__);} while(0)
+#else
+#define mutex_debug(x, ...)
+#endif
+
+#ifdef KERNEL
+#define mutex_pause kern_pause
+#else
+#define mutex_pause()
+#endif
+
+/* This implements a two-phase mutex which optionally allows a degree for
+   spinning when aquiring the lock. A limitation of this is that the spinning
+   is only for an uncondended lock, and the waiter does not enter the queue of
+   waiters until the second phase (context suspend). As such, waiters for a
+   mutex are not strictly processed in FIFO order.
+
+   Linux uses an MCS lock for the first phase, but the unqueueing (removal)
+   operation is complex (yet necessary in order to migrate waiters into the
+   queue of suspended tasks). It may be worth implementing something similar
+   here, perhaps moreso for the lessened bus contention than the FIFO
+   property. */
+
+static inline void mutex_acquired(mutex m, context ctx)
+{
+    assert(!m->turn);
+    m->turn = ctx;
+    ctx->waiting_on = 0;
+}
+
+static inline boolean mutex_cas_take(mutex m, context ctx)
+{
+    if (compare_and_swap_32(&m->count, 0, 1)) {
+        mutex_acquired(m, ctx);
+        return true;
+    }
+    return false;
+}
+
+static inline boolean mutex_lock_internal(mutex m, boolean wait)
+{
+    cpuinfo ci = current_cpu();
+    context ctx = get_current_context(ci);
+
+    mutex_debug("cpu %d, mutex %p, wait %d, ra %p\n", ci->id, m, wait,
+                __builtin_return_address(0));
+    mutex_debug("   ctx %p, turn %p, count %d\n", ctx, m->turn, m->count);
+
+    /* not preemptable (but could become so if needed) */
+    if (m->turn == ctx)
+        halt("%s: lock already held - cpu %d, mutex %p, ctx %p, ra %p\n", __func__,
+             ci->id, m, ctx, __builtin_return_address(0));
+
+    if (!wait)
+        return mutex_cas_take(m, ctx);
+
+    u64 spins_remain = m->spin_iterations;
+    while (spins_remain-- > 0) {
+        if (mutex_cas_take(m, ctx))
+            return true;
+        mutex_pause();
+    }
+
+    if (fetch_and_add_32(&m->count, 1) == 0) {
+        mutex_acquired(m, ctx);
+        return true;
+    }
+
+    /* race covered by dequeue loop in unlock */
+    assert(!frame_is_full(ctx->frame));
+
+    mutex_debug("ctx %p about to wait, count %d\n", ctx, m->count);
+    ctx->waiting_on = m;
+    while (!enqueue(m->waiters, ctx)) {
+        mutex_pause();      /* XXX timeout */
+    }
```

Comment:
In a condition where the waiter queue is full, this would present the same problem as a spinlock, i.e. the mutex-acquiring context, by spinning indefinitely, could prevent a suspended mutex-holding context from being resumed and releasing the mutex.
I wonder if we should add a struct list to struct context and replace the queue here with a linked list (and maybe replace the `dependents` vector in struct pending_fault with a list as well), or perhaps use a resizable vector instead of a queue. But I realize we would need to protect the list (or the vector) with an additional lock internal to the mutex (similar to the faulting_lock used to handle page faults).

Reply:
Ok, I think I have a solution. In the latest commit, the waiters queue is replaced by a lock-free singly-linked list (based on the one described in the MCS lock papers) composed of waiting contexts.

For another PR, I'll look at adding generic support for such a list - e.g. to use for pending_fault dependents. There are probably a number of places in the kernel that could benefit from such a lock-free list.